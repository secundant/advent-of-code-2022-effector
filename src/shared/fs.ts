import { readFile } from 'node:fs/promises';

export const readRelativeFile = (meta: ImportMeta, path: string) =>
  readFile(new URL(path, meta.url).pathname, 'utf-8');
