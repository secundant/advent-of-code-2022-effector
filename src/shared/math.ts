export const sum = (values: number[]) => values.reduce((acc, value) => acc + value, 0);
export const max = (values: number[]) => Math.max(...values);
