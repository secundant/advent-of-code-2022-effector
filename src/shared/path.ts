import { resolve } from 'node:path';

export const metaDirname = (meta: ImportMeta) => new URL('.', meta.url).pathname;
export const metaResolve = (meta: ImportMeta, ...paths: string[]) =>
  resolve(metaDirname(meta), ...paths);
