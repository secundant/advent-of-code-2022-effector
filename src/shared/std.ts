export type Comparator<T> = (left: T, right: T) => number;

// eslint-disable-next-line @typescript-eslint/require-array-sort-compare
export const sort = <T>(list: T[], comparator?: Comparator<T>) => Array.from(list).sort(comparator);

export const numericComparator = (a: number, b: number) => a - b;

export const last = <T>(list: T[]) => list.at(-1);
export const uniq = <T>(list: T[]) => Array.from(new Set(list));

export const divideString = (input: string, length: number) =>
  Array.from({ length }, (_, index) => input.slice(index * length, (index + 1) * length));

export const isUpperCase = (char: string) => char === char.toUpperCase();

export const getCharCode = (char: string) => char.charCodeAt(0);
