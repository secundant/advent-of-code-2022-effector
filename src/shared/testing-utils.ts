import { readFile } from 'node:fs/promises';

export const createTestContext = (meta: ImportMeta) => {
  const loadFile = (name = './data.txt') => readFile(new URL(name, meta.url).pathname, 'utf-8');

  return {
    loadFile
  };
};
