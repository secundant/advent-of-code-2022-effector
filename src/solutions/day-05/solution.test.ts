import { fork } from 'effector';
import { describe, expect, test } from 'vitest';
import { max } from '@/shared/math';
import { fromLength, length } from '@/shared/std';
import { getDayData } from '@/shared/testing-utils';
import {
  $commands,
  $reorderedMask,
  $reorderedStacks,
  $reorderedWithNewCrateMask,
  $stacks
} from '@/solutions/day-05/solution';

const exampleData = `    [D]
[N] [C]
[Z] [M] [P]
 1   2   3

move 1 from 2 to 1
move 3 from 1 to 3
move 2 from 2 to 1
move 1 from 1 to 2`;

const stacksIsEq = (actual: string[][], expected: string[][]) =>
  expect(actual).toEqual(expected.map(stack => stack.map(name => `[${name}]`)));

describe('Day 04', async () => {
  const createScope = (data: string) => {
    const [stacksInputWithLegend, commandsInput] = data.split('\n\n');
    const stacksInputs = stacksInputWithLegend.split('\n').slice(0, -1);
    const stacksLength = (max(stacksInputs.map(length)) + 1) / 4;
    const stacks = fromLength(stacksLength, x =>
      fromLength(stacksInputs.length, y =>
        stacksInputs[stacksInputs.length - 1 - y].slice(x * 4, x * 4 + 3).trim()
      ).filter(Boolean)
    );

    const commands = commandsInput
      .split('\n')
      .map(line => line.trim())
      .filter(Boolean)
      .map(line => {
        const [_, size, from, to] = line.match(/move ([0-9]*) from ([0-9]*) to ([0-9]*)/)!;

        return {
          size: Number.parseInt(size, 10),
          from: Number.parseInt(from, 10) - 1,
          to: Number.parseInt(to, 10) - 1
        };
      });

    return fork({
      values: [
        [$stacks, stacks],
        [$commands, commands]
      ]
    });
  };

  test('01 - should pass example', () => {
    const scope = createScope(exampleData);

    expect(scope.getState($reorderedMask)).toEqual('CMZ');
    stacksIsEq(scope.getState($reorderedStacks), [['C'], ['M'], ['P', 'D', 'N', 'Z']]);
  });

  test('01 - After the rearrangement procedure completes, what crate ends up on top of each stack?', async () => {
    const scope = createScope(await getDayData(5));

    expect(scope.getState($reorderedStacks)).toMatchSnapshot();
    expect(scope.getState($reorderedMask)).toMatchSnapshot();
  });

  test('02 - should pass example', () => {
    const scope = createScope(exampleData);

    expect(scope.getState($reorderedWithNewCrateMask)).toEqual('MCD');
  });

  test('02 - After the rearrangement procedure completes, what crate ends up on top of each stack?', async () => {
    const scope = createScope(await getDayData(5));

    expect(scope.getState($reorderedWithNewCrateMask)).toMatchSnapshot();
  });
});
