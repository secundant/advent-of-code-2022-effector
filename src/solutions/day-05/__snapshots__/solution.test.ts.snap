// Vitest Snapshot v1

exports[`Day 04 > 01 - After the rearrangement procedure completes, what crate ends up on top of each stack? 1`] = `
[
  [
    "[C]",
    "[M]",
    "[N]",
    "[H]",
  ],
  [
    "[J]",
    "[N]",
  ],
  [
    "[S]",
  ],
  [
    "[L]",
    "[L]",
    "[V]",
    "[N]",
  ],
  [
    "[C]",
    "[L]",
    "[D]",
    "[M]",
  ],
  [
    "[M]",
    "[R]",
    "[B]",
    "[V]",
    "[F]",
    "[Q]",
    "[W]",
    "[T]",
    "[Q]",
    "[J]",
    "[P]",
    "[L]",
    "[W]",
    "[H]",
    "[B]",
    "[R]",
    "[T]",
    "[G]",
    "[M]",
    "[H]",
    "[D]",
    "[T]",
  ],
  [
    "[D]",
    "[G]",
    "[Q]",
    "[F]",
    "[R]",
    "[Z]",
    "[G]",
    "[D]",
    "[P]",
    "[G]",
    "[H]",
    "[P]",
    "[L]",
    "[L]",
  ],
  [
    "[M]",
    "[C]",
    "[H]",
  ],
  [
    "[R]",
    "[Q]",
  ],
]
`;

exports[`Day 04 > 01 - After the rearrangement procedure completes, what crate ends up on top of each stack? 2`] = `"HNSNMTLHQ"`;

exports[`Day 04 > 02 - After the rearrangement procedure completes, what crate ends up on top of each stack? 1`] = `"RNLFDJMCT"`;
